rs_server/s3_storage_handler/s3_storage_handler.py

```diff
@@ -17,243 +24,719 @@
+
+        get_s3_data(s3_url):
+            Parses S3 URL to extract bucket, prefix, and file.
+
+        check_bucket_access(self, bucket):
+            Checks the accessibility of an S3 bucket.
+    """
+
+    def __init__(self, access_key_id, secret_access_key, endpoint_url, region_name):
+        """Initialize the S3StorageHandler instance.
+
+        Args:
+            access_key_id (str): The access key ID for S3 authentication.
+            secret_access_key (str): The secret access key for S3 authentication.
+            endpoint_url (str): The endpoint URL for the S3 service.
+            region_name (str): The region name.
+
+        Raises:
+            RuntimeError: If the connection to the S3 storage cannot be established.
+        """
+        self.logger = logging.getLogger("s3_storage_handler_test")
+        self.logger.setLevel(logging.DEBUG)
+        self.logger.handlers = []
+        self.logger.addHandler(logging.StreamHandler(sys.stdout))
+        self.logger.info("S3StorageHandler created !")
+
+        self.s3_client_mutex = Lock()
+        self.access_key_id = access_key_id
+        self.secret_access_key = secret_access_key
+        self.endpoint_url = endpoint_url
+        self.region_name = region_name
+        self.s3_client: boto3.client = None
+        if not self.connect_s3():
+            raise RuntimeError("The connection to s3 storage could not be made")
+
+    # get the s3 handler
+    def __get_s3_client(self, access_key_id, secret_access_key, endpoint_url, region_name):
+        """Retrieve or creates an S3 client instance.
+
+        Args:
+            access_key_id (str): The access key ID for S3 authentication.
+            secret_access_key (str): The secret access key for S3 authentication.
+            endpoint_url (str): The endpoint URL for the S3 service.
+            region_name (str): The region name.
+
+        Returns:
+            boto3.client: An S3 client instance.
+        """
+        # This mutex is needed in case of more threads accessing at the same time this function
+        with self.s3_client_mutex:
+            if self.s3_client:
+                return self.s3_client
+            client_config = botocore.config.Config(
+                max_pool_connections=100,
+                retries={"total_max_attempts": 10},
+            )
+            try:
+                s3_client = boto3.client(
+                    "s3",
+                    aws_access_key_id=access_key_id,
+                    aws_secret_access_key=secret_access_key,
+                    endpoint_url=endpoint_url,
+                    region_name=region_name,
+                    config=client_config,
+                )
+            except ClientError as e:
+                if e.response["Error"]["Code"] == "EntityAlreadyExists":
+                    self.logger.error("This clent already exists")
+                else:
+                    self.logger.error("Unexpected error")
+                return None
```

Comment:
In modern code, in case of error, we should not return an boolean but raise an Exception. So everyone does not have to check for return values. I've done it here: https://github.com/RS-PYTHON/rs-server/commit/11c017e1fb155a52e83d361febbd9676dd471088

Note that instead of
```
            except ClientError as e:
                if e.response["Error"]["Code"] == "EntityAlreadyExists":
                    self.logger.error("This clent already exists")
                else:
                    self.logger.error("Unexpected error")
```
You can do
```
            except ClientError as e:
                if e.response["Error"]["Code"] == "EntityAlreadyExists":
                    raise RuntimeError("This clent already exists") from e
                else:
                    raise e
```
So you always raise an exception and keep the original exception info.

Reply:
Merged with the branch you added the exceptions